lib/UndoTruncateToOddIntegerPass.cpp

```diff
@@ -160,32 +206,50 @@ bool UndoTruncatedSwitchConditionPass::runOnModule(Module &M) {
   zombies_.reset();
   i32_ = Type::getInt32Ty(M.getContext());
 
-  for (auto SI : WorkList) {
-    auto Cond = SI->getCondition();
-
-    auto *widened = ZeroExtend(Cond);
-    SI->setCondition(widened);
-    Changed = true;
-
-    for (auto Cases : SI->cases()) {
-      // The original value of the case.
-      auto V = Cases.getCaseValue()->getZExtValue();
+  while (!WorkList.empty()) {
+    auto inst = WorkList.back();
+    WorkList.pop_back();
 
-      // A new value for the case with the correct type.
-      auto CI = dyn_cast<ConstantInt>(ConstantInt::get(i32_, V));
+    auto extended = ZeroExtend(inst);
+    if (extended && extended != inst) {
+      Changed = true;
 
-      // And we replace the old value.
-      Cases.setValue(CI);
+      for (auto user : inst->users()) {
+        if (auto user_inst = dyn_cast<Instruction>(user)) {
+          WorkList.push_back(user_inst);
+        }
+      }
     }
   }
 
-  // Remove the zombies if we can.  We expect to.  They are ordered from
-  // combinations down to their supporting values.
+  // Remove the zombies if we can.  We expect to. Generally this can be done in
+  // 1 or 2 passes.
```

Comment:
I couldn't come with a scenario where we *wouldn't* be able to remove all zombie instructions (i.e. no instruction that is replaced should ever still have users).

Since we're doing the replacement recursively, can't we just build a stack of instructions to remove and then do them all bottom up (in one pass)? I just hacked this into your changes and they worked fine for all existing tests. If not, we're missing test coverage for the case where this doesn't work.

Reply:
Hmm I think I must have tested it wrong in development because UniqueVector indexing starts at 1.